Step3 구간추가 기능 리뷰 요청드립니다.

hyunhyun opened the pull request:
step3 구간 추가 리뷰요청드립니다. 
질문이 있는데, Line(Entity)안에서 LineStations(Embeddable)을 변경하는 메소드(addLineSection)으로 해서 기능 정상 되는걸 확인했는데요. 처음에는 Service에서 @Transactional로 설정한 메소드에서 LineStations 새인스턴스를 생성하고, line에서 가지고 있던 기존 건은 지우고, line에 lineStations를 새 객체로 변경하려했는데 아래와 같은 오류가 떠서 변경 했었습니다. 

Embedded 타입 때문인거 같은데, 새 객체로 할당하고, 기존 거의 list clear, 또는 repsitory 통해서 delete해도 아래가 계속 떳었었는데...
Embedded 타입이면 사용하는 엔티티 객체에서 호출해야지만 되는건가요...? 많이 찾아봤는데 답을 정확히 모르겠어서 질문드립니다..

"A collection with cascade=”all-delete-orphan” was no longer referenced by the owning entity instance".

hyunhyun committed:
fix : line put, delete api response 타입 명세서 기준으로 수정

hyunhyun committed:
feat : 구간 추가 기능 구현

oper912 commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -15,27 +17,43 @@
+    private final LineRepository lineRepository;
+    private final LineStationRepository lineStationRepository;
+    private final StationService stationService;
+    private final LineStationService lineStationService;
```
Service가 Service에 의존하는 경우 (동일한 레벨의 계층)
아래와 같은 문제들이 발생할 수 있어요 

- 생성자 주입 방식을 사용하고 있기 때문에 Service 들 간의 순환참조 문제가 발생할 가능성이 열려있는 구조
- 레이어간에 의존관계가 단방향으로 흐르지 않아서 확장성이나 유지보수 측면에서 계속 해서 힘들어짐 


추가로 이런 문제들을 해결하기 위해 Facade패턴을 사용(Facade layer 추가) 하거나
Service들간에도 QueryService + CommandService를 추가하는 등의 구조로 해결을 하기도 하는데 현재 미션에서는 계층간의 구분만 확실히 해주어도 좋을 것 같습니다

hyunhyun replied:
service에서 순환하면서 서로 참조하지 않도록 변경했습니다.

oper912 commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -15,27 +17,43 @@
+
+        lineStationService.addSection(
+                new SectionRequest(lineRequest.getUpStationId(), lineRequest.getDownStationId(),
+                        lineRequest.getDistance()), persistLine.getId());
```
Line에서 구간의 영속성전이(cascade) 옵션이 ALL로 되어있으므로 Line에서 자식들의 라이프 사이클을 관리할 수 있을 것 같은데요
Line 먼저 저장후에 구간을 추가하지말고 구간을 모두 세팅하고 Line을 save 해보면 어떨까요??

예시 코드.. 설계한 구조가 달라서 동일하게 적용은 못하지만 참고만 부탁드립니다
```
    @Transactional
    public LineResponse saveLine(final LineRequest request) {
        Station upStation = stationService.findById(request.getUpStationId());
        Station downStation = stationService.findById(request.getDownStationId());
        Line persistLine = lineRepository.save(
                new Line(request.getName(), request.getColor(), upStation, downStation, request.getDistance(), request.getAdditionalFare()));
        return LineResponse.of(persistLine);
    }
```

hyunhyun replied:
넵 cascade를 활용하는 방면으로 수정하였습니다.

oper912 commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -15,27 +17,43 @@
-        return LineResponse.from(line);
+        LineResponse lineResponse = LineResponse.from(line);
+
+        List<LineStation> byLineId = lineStationRepository.findByLineId(id);
+        for (LineStation lineStation : byLineId) {
+            lineResponse.addStationResponses(stationService.findStation(lineStation.getStationId()));
+        }
+        return lineResponse;
```
확인을 못해봐서 여쭈어보는데 이렇게 하면 역의 정렬순서는 보장되는 걸까요?? 

먼저 StationsResponses를 모두 획득한 뒤에 LineResponse를 생성해주는게 더 좋아보이네요

oper912 replied:
추가로 쿼리가 매우 많이 발생할거같네요 ㅠ

oper912 commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -45,9 +63,9 @@ public void deleteLine(Long id) {
 
     @Transactional
-    public LineResponse updateLine(Long id, LineRequest lineRequest) throws NotFoundException {
+    public void updateLine(Long id, LineRequest lineRequest) throws NotFoundException {
```
클라이언트에서 별도로 예외를 처리해줄게 아니라면 예외처리의 책임을 전가하지않는 게 더 좋아보입니다
```suggestion
    public void updateLine(Long id, LineRequest lineRequest) {
```

oper912 replied:
@ControllerAdvice or @RestControllerAdvice 를 활용해서 별도로 전역으로 예외를 핸들링 해보면 어떨까요??

oper912 replied:
예시 코드

```
@RestControllerAdvice
public class GlobalExceptionHandler {

    private static final Logger LOG = LoggerFactory.getLogger(GlobalExceptionHandler.class);

    @ExceptionHandler(value = IllegalArgumentException.class)
    public ResponseEntity handleIllegalArgsException(IllegalArgumentException e) {
        LOG.error("GlobalExceptionHandler.handleIllegalArgsException : ", e);
        return ResponseEntity.badRequest().build();
    }
    
```

hyunhyun replied:
GlobalExceptionHandler에서 처리하도록 변경하였습니다.

oper912 commented on src/main/java/nextstep/subway/application/LineStationService.java:
```diff
@@ -0,0 +1,43 @@
+        this.lineRepository = lineRepository;
+    }
+
+    @Transactional
+    public LineResponse addSection(SectionRequest sectionRequest, Long lineId) {
+
+        Line line = lineRepository.findById(lineId).get();
+        line.addLineStation(sectionRequest);
+        LineStations lineStations = line.getLineStations();
+
+        LineResponse lineResponse = LineResponse.from(line);
+        for (LineStation lineStation : lineStations.getLineStations()) {
+            Station station = stationRepository.findById(lineStation.getStationId()).get();
+            lineResponse.addStationResponses(StationResponse.from(station));
+        }
+        return lineResponse;
+    }
```
꼭 필요한 메서드인지 고민해보면 좋을 것 같습니다

hyunhyun replied:
리팩토링하면서 불필요해져 삭제하였습니다~!

oper912 commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -20,14 +22,23 @@ public class Line extends BaseEntity {
     private String color;
 
-    protected Line() {
+    @Embedded
+    LineStations lineStations = new LineStations();
+
+    public Line() {
```
질문주신 부분인거 같은데 사실.. 제가 이해를 잘 못했어요 😅 
글로만 설명해주시는거보다는 코드를 복붙 or 스샷으로 보여주시거나
문제가 되는 상황의 코드를 별도 브랜치로 관리해서 전달해주셨으면 더 좋았을 것 같아요 

추가로 의견을 말씀드리면
1. 구간을 추가하는데 왜 기존의 구간들을 모두 제거하고 새로운 구간들로 대체하는지부터가 문제가 있어보입니다
2. 임베디드 타입 활용이 어려우시면 먼저 구현을 완료하시구 이후에 리팩토링 하시는걸 추천드려요

oper912 replied:
또한 JPA가 익숙하지않고 미션이 어려우시다면 다른 리뷰이 분들의 코드를 참고하시는 것도 매우 좋은 방법입니다!

hyunhyun replied:
jpa가 익숙하지 않아서인지 예상대로 잘안되서 많이 어려웠던거 같습니다. 조언 감사합니다!

oper912 commented on src/main/java/nextstep/subway/domain/LineStationRepository.java:
```diff
@@ -0,0 +1,8 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface LineStationRepository extends JpaRepository<LineStation, Long> {
+    List<LineStation> findByLineId(long lineId);
```
아래 처럼 변경해볼 수도 있겠네요
```suggestion
    List<LineStation> findAllByLine(Line line);
```

hyunhyun replied:
리팩토링하면서 해당 함수 미사용 메소드되어 삭제하였습니다.

oper912 commented on src/main/java/nextstep/subway/domain/LineStations.java:
```diff
@@ -0,0 +1,164 @@
+@Embeddable
+public class LineStations {
+
+    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
```
```suggestion
    @OneToMany(mappedBy = "line", cascade = CascadeType.ALL, orphanRemoval = true)
```

oper912 replied:
양방향 연관관계일 경우 연관관계의 주인이 아닌곳에서는 mappedBy 를 활용해주셔야 될 것 같습니다

oper912 commented on src/main/java/nextstep/subway/dto/StationResponse.java:
```diff
@@ -32,11 +38,15 @@ public String getName() {
-    public LocalDateTime getCreatedDate() {
-        return createdDate;
-    }
+//    public LocalDateTime getCreatedDate() {
+//        return createdDate;
+//    }
+//
+//    public LocalDateTime getModifiedDate() {
+//        return modifiedDate;
+//    }
```
불필요한 코드는 제거!

hyunhyun replied:
넵!

oper912 commented on src/main/java/nextstep/subway/ui/LineController.java:
```diff
@@ -48,16 +53,32 @@ public ResponseEntity<LineResponse> readLine(@PathVariable long id) throws NotFo
+        return ResponseEntity.ok().build();
+    }
+
+    @PostMapping("/{lineId}/sections")
+    public ResponseEntity<LineResponse> addSection(@PathVariable Long lineId,
+                                                   @RequestBody SectionRequest sectionRequest) {
+        LineResponse lineResponse = sectionService.addSection(sectionRequest, lineId);
```
```suggestion
    @PostMapping("/{lineId}/sections")
    public ResponseEntity addSection(@PathVariable Long lineId,
                                                   @RequestBody SectionRequest sectionRequest) {
        LineResponse lineResponse = sectionService.addSection(sectionRequest, lineId);
```
![스크린샷 2022-06-16 오전 1 27 19](https://user-images.githubusercontent.com/31276206/173878172-1745c33d-6692-4526-8b7d-c14fd4d3e25b.png)

hyunhyun replied:
요구사항과 달랐군요... 반영하였습니다.

oper912 reviewed (changes requested):
안녕하세요 정현님 :)
3단계 진행해주신거에 대해 몇가지 코멘트 남겼는데 확인부탁드려요
고민 후에 리뷰 재요청 해주세요! 🙏

hyunhyun committed:
refactor : lineStation cascade로 등록 활용, lineResponse 변경

lineResponse : stations -> lineStations

hyunhyun committed:
refactor : 미사용 메소드 삭제,구간 추가 response 타입 변

LineController : addSection return 타입 변경
GlobalExceptionHandler : linecontroller exception 추가
Line : lineStationResponse 파라미터 생성자 lineStations에서 추가로 변경

oper912 reviewed (approved):
피드백 반영 잘해주셨네요 👍👍
바로 머지하도록 하겠습니다
마지막까지 화이팅입니다 💪
